### ISSUE TITLE
Shader parameters specified in the inspector are not included when exporting a project in headless mode

### ISSUE THREAD

**Issue by GullikX** [godotengine/godot#66842]
### Godot version

4.0.beta2.official.f8745f2f7

### System information

Void linux

### Issue description

When exporting a project using `--headless --export-pack`, any shader parameters/uniforms specified in the inspector are not included in the exported project. All shader parameters are reverted to their default values (or null).

### Steps to reproduce

Project setup:
1. Create a cube mesh.
2. Add a shader material to the cube with the following simple shader:
```
shader_type spatial;
render_mode unshaded;

uniform vec4 color: source_color = vec4(0.0, 1.0, 0.0, 0.0); // Green

void fragment() {
        ALBEDO = color.rgb;
}
```
3. In the inspector, set the 'color' shader parameter/uniform to be a red color.
4. When running the project from the editor, the cube is now red.

Export without `--headless` option:
```
rm -rf .godot
./Godot_v4.0-beta2_linux.x86_64 --editor --export-pack Linux/X11 export.zip
./Godot_v4.0-beta2_linux.x86_64 --main-pack export.zip
# Cube is red (correct behavior)
```

Export with `--headless` option:
```
rm -rf .godot
./Godot_v4.0-beta2_linux.x86_64 --editor --headless --export-pack Linux/X11 export_headless.zip
./Godot_v4.0-beta2_linux.x86_64 --main-pack export_headless.zip
# Cube is green (incorrect behavior)
```

The behavior can be reproduced on Godot 4.0 alpha 16 through Godot 4.0 beta 2.

### Minimal reproduction project

[ShaderParameterHeadless.zip](https://github.com/godotengine/godot/files/9700232/ShaderParameterHeadless.zip)

**Comment by GullikX** [comment]
Git bisect indicates that this bug was introduced in commit ef17c46.

**Comment by GullikX** [comment]
It seems like shader parameters get lost when scene files are converted to binary during export. A workaround is to disable 'convert text resources to binary' in the project settings.

**Comment by jayypluss** [comment]
I've had the same problem, I changed the default params of the shader to reflect what I wanted directly in the shader file. 

> It seems like shader parameters get lost when scene files are converted to binary during export. A workaround is to disable 'convert text resources to binary' in the project settings.

This option is under Editor -> Export, you need to activate the Advanced Settings switch.

**Comment by cgsdev0** [comment]
I spent a lot of time investigating this tonight

It seems like this code is problematic in headless mode:
https://github.com/godotengine/godot/blob/4dcf2c5431ea6894c268c384df0d81e7fd96c341/editor/export/editor_export_platform.cpp#L742-L773

My working theory (as a total godot codebase noob) is that when we instantiate the node in headless mode, certain things (i.e. shader parameters) are ignored since they're unnecessary. This makes sense when we're running the game, but not when we're exporting.

I plan to continue working on this issue, but I figured I'd leave some notes here for myself and anyone interested.

**Comment by cgsdev0** [comment]
I think I've narrowed down the problem, but I'm not really sure what a good impl for a fix looks like here.

https://github.com/godotengine/godot/blob/a7276f1ce0c2911216a2c4718efddab98ddffd8f/scene/resources/shader.cpp#L117-L121

This function gets called in the path of exporting now, which, in headless mode, ends up calling this function:
https://github.com/godotengine/godot/blob/a7276f1ce0c2911216a2c4718efddab98ddffd8f/servers/rendering/dummy/storage/material_storage.h#L69

Since the headless rendering server doesn't keep track of shader parameters, they get lost when we instantiate the tree.

**Comment by davidsmith1307** [comment]
Can confirm that the issue is still present in 4.0.3 and the workaround still works.

**Comment by RGonzalezTech** [comment]
This issue is still present in 4.1 and the workaround still solves it 👍 This was affecting my `GPUParticles3D` as well.

**Comment by tcoxon** [comment]
I have also encountered this in 4.1.1-stable. The workaround worked for me.

The weird thing in my case is that it was happening even with GUI exports. Seems like it affects exports from the editor if you've used `--headless` in the past (maybe if the first export was `--headless`?). Wiping `.godot/exported` and then re-exporting from the editor fixed that for me. (But of course I still can't use `--headless` to build my project.)

**Comment by akien-mga** [comment]
CC @godotengine/shaders 

Would be worth diffing the `.godot/exported` folder generated via `--headless` or via the editor GUI, to see if the difference can be identified.

**Comment by tcoxon** [comment]
> Would be worth diffing the `.godot/exported` folder generated via `--headless` or via the editor GUI, to see if the difference can be identified.

cgsdev0's comments are correct. It looks like headless exporting calls into the dummy rendering server to get the shader's parameter list, which is just a stub. Setting a breakpoint on the dummy render server methods verifies this. The question now seems to be what the fix should be, rather than what the issue is.

**Comment by Calinou** [comment]
I assume the RenderingDriverDummy needs to set per-instance parameters on the CPU, so that they can be retrieved later (even if they are never required for rendering).

**Comment by Polemus** [comment]
> FYI, I can confirm that this issue still exists with Godot 4.2, and the work-around still resolves the problem.

yup, 4.2 mono here, this is still an issue, and the workaround works as well

**Comment by QbieShay** [comment]
When folks say "the workaround still works", do they mean giving the shader the default params equal to the desired ones?

**Comment by PeteX** [comment]
@QbieShay no I mean disabling "convert text resources to binary" in the project settings.

**Comment by mghicks** [comment]
For me in 4.2.1, this issue occurs both in headless exports and from-editor exports. Disabling convert text resources to binary solves the missing shader parameter, enabling convert text to binary causes the missing param in both headless and from-editor export. I am very happy to have found this issue and the workaround after some months of wondering if this was an undocumented compatibility renderer limitation.

**Comment by Eole211** [comment]
> For me in 4.2.1, this issue occurs both in headless exports and from-editor exports. Disabling convert text resources to binary solves the missing shader parameter, enabling convert text to binary causes the missing param in both headless and from-editor export. I am very happy to have found this issue and the workaround after some months of wondering if this was an undocumented compatibility renderer limitation.

It was the same for me. Maybe it happens since I did an headless export for the first time.  I had to go to each of my custom shaders to make them unique each time I really needed them for an export. Finding this issue and the fix was really helpful.

**Comment by Krigu** [comment]
We are building in a (Linux) CI Pipeline in headless mode. After hours of frustration and wondering why it worked locally but not on the exported versions of iOS and web, we were also able to solve the issue with disabling the project setting "convert text resources to binary".

**Comment by Tichau** [comment]
I think it needs to be reopen.
This issue is still present with v4.3.stable.mono.official [77dcf97d8] (the workaround still solves it).

**Comment by clayjohn** [comment]
@jupiterbjy Do you have a new MRP for testing? (or @Tichau?)

I can't reproduce the issue with 4.3 using the MRP from the OP.

**Comment by jupiterbjy** [comment]
@clayjohn Not mre yet but have proof here which I'm making as part of company work:

![image](https://github.com/user-attachments/assets/762e932d-1afe-42f7-9cc8-0c0054714450)

So ground tile there is simple shader that has instance parameters for grid cell size, and as you see one with `--headless` this is changed backt to default value.

so TL;DR instance param still seems to be broken. I'll try making MRE in free time, sorry I can't directly show you code of this haha

**Comment by jonjondev** [comment]
Our team is currently experiencing this issue as of 4.3 binary release: https://github.com/CapsCollective/sunflower/pull/45

Setting `convert_text_resources_to_binary` provides a temporary workaround. Waiting for proper fix, however.

**Comment by clayjohn** [comment]
So it looks like both instance uniforms and global uniforms are missed when doing a headless export. It should not be too much work to add them to the dummy rendering server as well
